### PR TITLE
Add Health Checks for Multi-Site deployment to HA Docs

### LINKED ISSUE #33143: Add the Troubleshooting and Health checks guide to Keycloak
```[tasklist]
### Tasks
- [ ] Move the health check docs to Keycloak project HA guide to help with troubleshooting the overall health of the multi-site deployment
```

### PR DESCRIPTION
**Pull request by kami619**
Fixes https://github.com/keycloak/keycloak/issues/33143

This PR aims to improve the troubleshooting aspects of a Multi-Site deployment of Keycloak on a Kubernetes environment and helps assert the overall health status of various components involved in the deployment.

### CODE REVIEW

**Review by ahus1 — DISMISSED** [review]
Thank you for this guide. IMHO it would be sufficient to assume only Kubernetes in this deployment and not OpenShift, which would make it more approachable to the wider community. 

So I'd ask you to change "oc" to "kubectl" and replace "OpenShift" with "Kubernetes" except where we reference the prerequisite of the previous installation explicitly.

**Comment by kami619** [comment]
@ahus1 I have addressed your previous review comments, please take another look if that's enough.

**Review by ahus1 — COMMENTED** [review]
Thank you for he docs. I updated the formatting a bit and shortened some of the headings and sentences for better readability. Please let me know if that breaks any meaning. 

I'll ask Andy to a review of the guide before this will then be merged.

**Comment by ahus1** [comment]
@andymunro - we would like to include this guide in KC26. Can you please review this new text? Thanks!

**Comment by ahus1** [comment]
@andymunro - gentle reminder. Please review when you have the time. We'd like to merge it tomorrow the latest to finish this task which is in our sprint. If you don't have the time, We could merge this version and postpone a review until the RHBK release. Thanks!

**Comment by andymunro** [comment]
Thanks for the reminder, @ahus1.  I will check it now.

File: docs/guides/high-availability/health-checks-multi-site.adoc [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+<@tmpl.guide
+title="Health checks for multi-site deployments"
+summary="Validating the health of a multi-site deployment" >
+
+When running the <@links.ha id="introduction" /> in a Kubernetes environment,
+you should automate checks to see if everything is up and running as expected.
+
+This page provides an overview of URLs,
+Kubernetes resources and Healthcheck endpoints available to verify a multi-site setup of {project_name}.
```
Comment (andymunro) [inline docs/guides/high-availability/health-checks-multi-site.adoc:12]:
```suggestion
This page provides an overview of URLs,
Kubernetes resources, and Healthcheck endpoints available to verify a multi-site setup of {project_name}.
```

File: docs/guides/high-availability/health-checks-multi-site.adoc [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+== Overview
+
+A proactive monitoring strategy aims to detect and alert on issues before they impact users. It is the key for a highly resilient and highly available {project_name} application.
```
Comment (andymunro) [inline docs/guides/high-availability/health-checks-multi-site.adoc:16]:
```suggestion
A proactive monitoring strategy aims to detect and alert about issues before they impact users. This strategy is the key for a highly resilient and highly available {project_name} application.
```

File: docs/guides/high-availability/health-checks-multi-site.adoc [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+Health checks across various architectural components (like application health, load balancing, caching, and overall system status) are critical for:
```
Comment (andymunro) [inline docs/guides/high-availability/health-checks-multi-site.adoc:18]:
```suggestion
Health checks across various architectural components (such as application health, load balancing, caching, and overall system status) are critical for:
```

File: docs/guides/high-availability/health-checks-multi-site.adoc [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+Ensuring High Availability:: By verifying that all sites and the load balancer are operational, it helps guarantee that the system can handle requests even if one site goes down.
```
Comment (andymunro) [inline docs/guides/high-availability/health-checks-multi-site.adoc:20]:
```suggestion
Ensuring High Availability:: Verifying that all sites and the load balancer are operational is a key to ensure that  a system can handle requests even if one site goes down.
```

File: docs/guides/high-availability/health-checks-multi-site.adoc [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+Maintaining Performance:: Checking the health and distribution of the {jdgserver_name} cache ensures that {project_name} can maintain optimal performance by efficiently handling sessions and other temporary data.
+
+Operational Resilience:: By continuously monitoring the health of both {project_name} and its dependencies within the Kubernetes environment, the system can quickly identify and possibly auto-remediate issues, reducing downtime.
+
+== Prerequisites
+
+. https://kubernetes.io/docs/tasks/tools/#kubectl[Kubectl CLI is installed and configured].
+
+. Install https://jqlang.github.io/jq/download/[jq] if it is not available on your OS already.
```
Comment (andymunro) [inline docs/guides/high-availability/health-checks-multi-site.adoc:30]:
```suggestion
. Install https://jqlang.github.io/jq/download/[jq] if it is not available on your operating system already.
```

File: docs/guides/high-availability/health-checks-multi-site.adoc [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+== Overview
+
+A proactive monitoring strategy aims to detect and alert on issues before they impact users. It is the key for a highly resilient and highly available {project_name} application.
+
+Health checks across various architectural components (like application health, load balancing, caching, and overall system status) are critical for:
+
+Ensuring High Availability:: By verifying that all sites and the load balancer are operational, it helps guarantee that the system can handle requests even if one site goes down.
+
+Maintaining Performance:: Checking the health and distribution of the {jdgserver_name} cache ensures that {project_name} can maintain optimal performance by efficiently handling sessions and other temporary data.
```
Comment (andymunro) [inline docs/guides/high-availability/health-checks-multi-site.adoc:22]:
```suggestion
Maintaining performance:: Checking the health and distribution of the {jdgserver_name} cache ensures that {project_name} can maintain optimal performance by efficiently handling sessions and other temporary data.
```

File: docs/guides/high-availability/health-checks-multi-site.adoc [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+Operational Resilience:: By continuously monitoring the health of both {project_name} and its dependencies within the Kubernetes environment, the system can quickly identify and possibly auto-remediate issues, reducing downtime.
```
Comment (andymunro) [inline docs/guides/high-availability/health-checks-multi-site.adoc:24]:
```suggestion
Operational resilience:: By continuously monitoring the health of both {project_name} and its dependencies within the Kubernetes environment, the system can quickly identify and possibly auto-remediate issues, reducing downtime.
```

File: docs/guides/high-availability/health-checks-multi-site.adoc [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+
+== Prerequisites
+
+. https://kubernetes.io/docs/tasks/tools/#kubectl[Kubectl CLI is installed and configured].
+
+. Install https://jqlang.github.io/jq/download/[jq] if it is not available on your OS already.
+
+== Specific Health Checks
```
Comment (andymunro) [inline docs/guides/high-availability/health-checks-multi-site.adoc:32]:
```suggestion
== Specific health checks
```

File: docs/guides/high-availability/health-checks-multi-site.adoc [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+
+=== {project_name} Load Balancer and Sites
```
Comment (andymunro) [inline docs/guides/high-availability/health-checks-multi-site.adoc:34]:
```suggestion
=== {project_name} load balancer and sites
```

File: docs/guides/high-availability/health-checks-multi-site.adoc [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+
+Verifies the health of the {project_name} application through its load balancer and both primary and backup sites. This ensures that {project_name} is accessible and that the load balancing mechanism is functioning correctly across different geographical or network locations.
+
+This command returns the health status of the {project_name} application's connection to its configured database, thus confirming the reliability of db connections.
```
Comment (andymunro) [inline docs/guides/high-availability/health-checks-multi-site.adoc:38]:
```suggestion
This command returns the health status of the {project_name} application's connection to its configured database, thus confirming the reliability of database connections.
```

File: docs/guides/high-availability/health-checks-multi-site.adoc [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+This is only available on the management port, and not available from the external URL.
+In a Kubernetes setup, the sub-status `health/ready` is checked periodically to make the Pod as ready.
```
Comment (andymunro) [inline docs/guides/high-availability/health-checks-multi-site.adoc:40]:
```suggestion
This command is available only on the management port and not from the external URL.
In a Kubernetes setup, the sub-status `health/ready` is checked periodically to make the Pod as ready.
```

File: docs/guides/high-availability/health-checks-multi-site.adoc [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+curl -s https://keycloak:managementport/health
+----
+
+This command verifies the `lb-check` endpoint of the load balancer and ensures the {project_name} application cluster is up and running.
+[source,bash]
+----
+curl -s https://keycloak-load-balancer-url/lb-check
+----
+
+These commands will return the running status of the Site A and Site B of the {project_name} in a cross-site setup (be it Active/Passive or Active/Active).
```
Comment (andymunro) [inline docs/guides/high-availability/health-checks-multi-site.adoc:53]:
```suggestion
These commands will return the running status of the Site A and Site B of the {project_name} in a cross-site setup (whether it is Active/Passive or Active/Active).
```

File: docs/guides/high-availability/health-checks-multi-site.adoc [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+[source,bash]
+----
+curl -s https://keycloak_site_a_url/lb-check
+curl -s https://keycloak_site_b_url/lb-check
+----
+
+=== {jdgserver_name} Cache Health
+Check the health of the default cache manager and individual caches in an external {jdgserver_name} cluster.
+This is vital for {project_name} performance and reliability,
+as {jdgserver_name} is often used for distributed caching and session clustering in {project_name} deployments.
```
Comment (andymunro) [inline docs/guides/high-availability/health-checks-multi-site.adoc:64]:
```suggestion
=== {jdgserver_name} Cache health
Check the health of the default cache manager and individual caches in an external {jdgserver_name} cluster.
This check is vital for {project_name} performance and reliability,
as {jdgserver_name} is often used for distributed caching and session clustering in {project_name} deployments.
```

File: docs/guides/high-availability/health-checks-multi-site.adoc [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+
+This command returns the overall health of the {jdgserver_name} cache manager, this is useful as the Admin user doesn't need to provide user credentials to get the health status.
```
Comment (andymunro) [inline docs/guides/high-availability/health-checks-multi-site.adoc:66]:
```suggestion
This command returns the overall health of the {jdgserver_name} cache manager, which is useful as the Admin user does not need to provide user credentials to get the health status.
```

File: docs/guides/high-availability/health-checks-multi-site.adoc [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+[source,bash]
+----
+curl -s https://infinispan_rest_url/rest/v2/cache-managers/default/health/status
+----
+
+Whereas for these health checks, Admin user needs to provide the {jdgserver_name} user credentials as part of the request to peek into the overall health of the external {jdgserver_name} cluster caches.
```
Comment (andymunro) [inline docs/guides/high-availability/health-checks-multi-site.adoc:72]:
```suggestion
In contrast to the preceding health checks, the following health checks require the Admin user to provide the {jdgserver_name} user credentials as part of the request to peek into the overall health of the external {jdgserver_name} cluster caches.
```

File: docs/guides/high-availability/health-checks-multi-site.adoc [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+[source,bash]
+----
+curl -u <infinispan_user>:<infinispan_pwd> -s https://infinispan_rest_url/rest/v2/cache-managers/default/health \
+ | jq 'if .cluster_health.health_status == "HEALTHY" and (all(.cache_health[].status; . == "HEALTHY")) then "HEALTHY" else "UNHEALTHY" end'
+----
+
+The `jq` filter is a convenience to compute the overall health based on the individual cache health.
+Optionally you can run the above command without the `jq` filter to see the full details.
```
Comment (andymunro) [inline docs/guides/high-availability/health-checks-multi-site.adoc:80]:
```suggestion
You can also choose to run the above command without the `jq` filter to see the full details.
```

File: docs/guides/high-availability/health-checks-multi-site.adoc [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+
+=== {jdgserver_name} Cluster Distribution
```
Comment (andymunro) [inline docs/guides/high-availability/health-checks-multi-site.adoc:82]:
```suggestion
=== {jdgserver_name} Cluster distribution
```

File: docs/guides/high-availability/health-checks-multi-site.adoc [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+Assesses the distribution health of the {jdgserver_name} cluster, ensuring that the cluster's nodes are correctly distributing data. This step is essential for the scalability and fault tolerance of the caching layer.
+
+You can modify the `expectedCount 3` argument to match the total nodes in the cluster and validate if they are healthy or not.
+[source,bash]
+----
+curl <infinispan_user>:<infinispan_pwd> -s https://infinispan_rest_url/rest/v2/cluster\?action\=distribution \
+ | jq --argjson expectedCount 3 'if map(select(.node_addresses | length > 0)) | length == $expectedCount then "HEALTHY" else "UNHEALTHY" end'
+----
+
+=== Overall, {jdgserver_name} System Health
```
Comment (andymunro) [inline docs/guides/high-availability/health-checks-multi-site.adoc:92]:
```suggestion
=== Overall, {jdgserver_name} system health
```

File: docs/guides/high-availability/health-checks-multi-site.adoc [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+Uses `kubectl` CLI tool to query the health status of {jdgserver_name} clusters and the {project_name} service in the specified namespace. This comprehensive check ensures that all components of the {project_name} deployment are operational and correctly configured within the Kubernetes environment.
```
Comment (andymunro) [inline docs/guides/high-availability/health-checks-multi-site.adoc:93]:
```suggestion
Uses the `kubectl` CLI tool to query the health status of {jdgserver_name} clusters and the {project_name} service in the specified namespace. This comprehensive check ensures that all components of the {project_name} deployment are operational and correctly configured within the Kubernetes environment.
```

File: docs/guides/high-availability/health-checks-multi-site.adoc [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+
+[source,bash]
+----
+kubectl get infinispan -n <NAMESPACE> -o json  \
+| jq '.items[].status.conditions' \
+| jq 'map({(.type): .status})' \
+| jq 'reduce .[] as $item ([]; . + [keys[] | select($item[.] != "True")]) | if length == 0 then "HEALTHY" else "UNHEALTHY: " + (join(", ")) end'
+----
+
+=== {project_name} Readiness in Kubernetes
```
Comment (andymunro) [inline docs/guides/high-availability/health-checks-multi-site.adoc:103]:
```suggestion
=== {project_name} readiness in Kubernetes
```

**Review by andymunro — CHANGES_REQUESTED** [review]
Thanks, @kami619. Well written.  Just a few suggestions.

**Comment by kami619** [comment]
@ahus1 I think this should be all set now, after @andymunro's review and feedback.

**Review by ahus1 — DISMISSED** [review]

**Review by ahus1 — APPROVED** [review]

**Pull request merged**